uma-sdk/src/commonMain/kotlin/me/uma/protocol/Invoice.kt

```
@@ -0,0 +1,275 @@
+    val requiredPayerData: CounterPartyDataOptions? = null,
+
+    /** UmaVersion is a list of UMA versions that the VASP supports for this transaction. It should be
+    * containing the lowest minor version of each major version it supported, separated by commas.
+    */
+    val umaVersion: String,
+
+    /** CommentCharsAllowed is the number of characters that the sender can include in the comment field of the pay request. */
+    val commentCharsAllowed: Int? = null,
+
+    /** The sender's UMA address. If this field presents, the UMA invoice should directly go to the sending VASP instead of showing in other formats. */
+    val senderUma: String? = null,
+
+    /** The maximum number of the invoice can be paid */
+    val invoiceLimit: Int? = null,
+
+    /** YC status of the receiver, default is verified. */
+    val kycStatus: KycStatus? = null,
+
+    /** The callback url that the sender should send the PayRequest to. */
+    val callback: String,
+
+    /** The signature of the UMA invoice */
+    val signature: ByteArray? = null,
+) : TLVCodeable {
+
+    override fun toTLV() = mutableListOf<ByteArray>()
+            .putString(0, receiverUma)
+            .putString(1, invoiceUUID)
+            .putNumber(2, amount)
+            .putTLVCodeable(3, receivingCurrency)
+            .putNumber(4, expiration)
+            .putBoolean(5, isSubjectToTravelRule)
+            .putByteCodeable(6, requiredPayerData?.let(::InvoiceCounterPartyDataOptions))
+            .putString(7, umaVersion)
+            .putNumber(8, commentCharsAllowed)
+            .putString(9, senderUma)
+            .putNumber(10, invoiceLimit)
+            .putByteCodeable(11, kycStatus?.let(::InvoiceKycStatus))
+            .putString(12, callback)
+            .putByteArray(100, signature)
+            .array()
+
+    fun toBech32(): String = Bech32.encode(Bech32.Encoding.BECH32, UMA_BECH32_PREFIX, this.toTLV())
+
+    companion object {
+        fun fromTLV(bytes: ByteArray): Invoice {
```

If a non-optional field is not presented, we should throw an error here.